Add retention subsystem (#55125)
zephyrproject-rtos/zephyr: doc/services/retention/index.rst

.. _retention_api:

Retention System
################

> The retention system provides an API which allows applications to read/write
> data from/to memory areas or devices which retain the data whilst the device is
> powered. This is designed for sharing information between different
> applications or for a single application when a device reboots without losing
> state information, the contents of the data should not persist in the event of
> a power failure (or possibly with some low power modes on some devices) nor
> should it be stored to a non-volatile backing store such as :ref:`flash_api`,
> :ref:`eeprom_api` or battery-backed RAM.

Reviewer: ```suggestion
The retention system provides an API which allows applications to read and write
data from and to memory areas or devices that retain the data while the device is
powered. This allows for sharing information between different
applications or within a single application without losing
state information when a device reboots .
The stored data should not persist in the event of
a power failure (or during some low-power modes on some devices) nor
should it be stored to a non-volatile storage like :ref:`flash_api`,
:ref:`eeprom_api`, or battery-backed RAM.
```

> The retention system builds on top of the retained data driver, and adds

Reviewer 2: Please link to docs that describe `retained data driver`.

Reviewer 2: What I meant was to put the link in the rst.

Reviewer 2: Is my understanding correct that the subsystem is only might not necessarily be compatible with any retained mem driver? For example, this would not work with https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/retained_mem/retained_mem_nrf_gpregret.c, correct?

Author: It would work fine with it. You can even see the test here https://github.com/zephyrproject-rtos/zephyr/pull/55125/commits/d7d180b17baf480e3cfd426a98e643acecc88dfd which runs on nRF52

additional software-level features to it for ensuring the validity of data.
Optionally, a magic header can be used to check if the front of
the retained data memory section contains this specific value, and an optional
checksum (1, 2 or 4-bytes in size) of the stored data can be appended to the
end of the data. Additionally, the retention system API allows partitioning of
retained data sections into multiple distinct areas, for example a 64-byte
retained data area could be split up into 4 bytes for a boot mode, 16 bytes
for a timestamp, 44 bytes for a last log message - all of these sections can
be accessed or updated independently. The prefix and checksum can be set
per-instance via device tree configuration.

Device tree setup
*****************

To use the retention system, a retained data driver must be setup for the board

> you are using, there is a zephyr driver which can be used which will use some

Reviewer 2: Which zephyr driver? 

From my understanding of the code this must be `retained_mem_zephyr_ram.c` driver. Correct? If that is correct, please state this.

Author: No, any retained memory driver.
